flexible-event/flexible_event_privacy.py

```diff
@@ -0,0 +1,167 @@
+
+    Args:
+    x: the input value.
+
+    Returns:
+      The binary entropy function at x.
+    """
+    if x == 0 or x == 1:
+        return 0
+    else:
+        return - x * log(x, 2) - (1 - x) * log(1 - x, 2)
+
+
+def flip_probability_dp(num_states: int, epsilon: float) -> float:
+    """Returns the flip probability to satisfy epsilon differential privacy.
+
+       Uses the k-RR privacy mechanism.
+    """
+    return num_states / (num_states + exp(epsilon) - 1)
+
+
+def capacity_q_ary_symmetric_channel(log2_q: float,
+                                     flip_probability: float) -> float:
+    """Computes the capacity of the q-ary symmetric channel.
+
+    Args:
+      log2_q: the logarithm to base 2 of the alphabet size.
+      flip_probability: the channel keeps the input the same with probability
+        1 - flip_probability, and flips the input to one of the other q - 1
+        symbols (uniformly) with the remaining probability of flip_probability.
+
+    Returns:
+      The capacity of the q-ary symmetric channel for given flip_probability.
+        In general, the capacity is defined as the maximum, over all input
+        distributions, of the mutual information between the input and output of
+        the channel. In the special case of the q-ary symmetric channel, a
+        closed-form expression is known, which we use here.
+    """
+    return (log2_q - h(flip_probability) - flip_probability * log(2**log2_q - 1, 2))
+
+
+def max_information_gain(num_states: int, epsilon: float):
+    """Returns the maximum information for a source using the flexible event API."""
+
+    flip_prob = flip_probability_dp(num_states, epsilon)
+    return capacity_q_ary_symmetric_channel(log(num_states, 2),
+                                            flip_prob*(num_states-1)/num_states)
+
+
+def get_config(json: dict) -> ApiConfig:
+    max_event_level_reports = json.get('max_event_level_reports', 20)
+    per_trigger_data_configs = []
+    for spec in json['trigger_specs']:
+        num_data_types = len(spec['trigger_data'])
+        num_windows = len(spec['event_report_windows']['end_times'])
+        num_buckets = len(spec['summary_buckets']
+                          ) if 'summary_buckets' in spec else 20
+        per_trigger_data_configs.extend(
+            [(num_windows, num_buckets)] * num_data_types)
+
+    return ApiConfig(max_event_level_reports, per_trigger_data_configs)
+
+
+if __name__ == "__main__":
+    DESCRIPTION = '''\
+  flexible_event_privacy.py is a utility to ingest configurations for the flexible
+  event-level reports in the Attribution Reporting API. It reads a JSON from stdin
+  matching the input in the request header Attribution-Reporting-Register-Source.
+
+  It optionally also accepts windows and per-trigger-data summary buckets from command
+  line arguments.
+
+  Note: JSON input is not completely validated. It is minimally processed to count the
+  number of windows and buckets per trigger spec.
```

Comment:
Change request: Add example usage / example call. This can be a follow-up, and does not need to live here in this file, but would be very helpful.

Ideally that example would include:
* One usage without the optional arguments (and the source header for context) + The corresponding ouput
* One usage with the optional arguments (and the source header for context) + The corresponding ouput

I can get something started.

Reply:
Ack